test/utils/test_github_utils.py

```diff
@@ -0,0 +1,147 @@
+                                              'mode': github_utils._FILE_MODE,
+                                              'type': 'blob',
+                                              'content': ''}]}),
+                                _USERNAME, _PASSWORD),
+                      mock.call(
+                          requests.post, _POST_TREE_URL,
+                          json.dumps({'base_tree': 'main_sha',
+                                      'tree':
+                                      [{'path': 'subdir',
+                                        'mode': github_utils._TREE_MODE,
+                                        'type': 'tree',
+                                        'sha': 'fake-tree-sha'},
+                                       {'path': 'hello.txt',
+                                        'mode': github_utils._FILE_MODE,
+                                        'type': 'blob',
+                                        'content': 'hello local\n'}]}),
+                          _USERNAME, _PASSWORD),
+                      mock.call(
+                          requests.post, _POST_TREE_URL,
+                          json.dumps(
+                              {'tree':
+                               [{'path': 'executable.py',
+                                 'mode': github_utils._EXEC_MODE,
+                                 'type': 'blob',
+                                 'content': '#!/usr/bin/env python\n'}]}),
+                          _USERNAME, _PASSWORD),
+                      mock.call(
+                          requests.post, _POST_TREE_URL,
+                          json.dumps({'base_tree': 'root_sha',
+                                      'tree':
+                                      [{'path': 'main',
+                                        'mode': github_utils._TREE_MODE,
+                                        'type': 'tree',
+                                        'sha': 'fake-tree-sha'},
+                                       {'path': 'util',
+                                        'mode': github_utils._TREE_MODE,
+                                        'type': 'tree',
+                                        'sha': 'fake-tree-sha'}]}),
+                          _USERNAME, _PASSWORD),
+                      mock.call(requests.post, _POST_COMMIT_URL,
+                                json.dumps({'tree': 'fake-tree-sha',
+                                            'parents': ['fake-commit-sha'],
+                                            'message': 'fake-message'}),
+                                _USERNAME, _PASSWORD),
+                      mock.call(requests.patch, _PATCH_REF_URL,
+                                json.dumps({'force': True,
+                                            'sha': 'fake-commit-sha'}),
+                                _USERNAME, _PASSWORD)]
+
+
+def _requests_side_effect(http_method, url, req_data=None, username=None,
```

Comment:
As discussed offline, I recommend not representing the remote tree, because it requires a substantial amount of complexity in this function -- ideally, the test should have little logic of its own.

Instead, can we mock `requests` and explicitly specify each response to each request, with no reference to the remote tree as a whole?